File: Rock/Web/UI/Controls/Pickers/DataViewPicker.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
using System;
using System.Collections.Generic;
using System.Data.Entity;
using System.Linq;
using System.Web;
using System.Web.UI.WebControls;

using Rock.Data;
using Rock.Model;
using Rock.Security;
using Rock.Web.Cache;

namespace Rock.Web.UI.Controls
{
    /// <summary>
    /// 
    /// </summary>
    [RockObsolete( "1.8" )]
    [Obsolete("Use DataViewItemPicker instead")]
    public class DataViewPicker : RockDropDownList
    {

        /// <summary>
        /// Gets or sets a value indicating if items should be recreated whenever a property value is changed.
        /// </summary>
        /// <value>
        ///   <c>true</c> if [automatic rebuild items]; otherwise, <c>false</c>.
        /// </value>
        public bool AutoLoadItems
        {
            get
            {
                return ViewState["AutoLoadItems"] as bool? ?? true;
            }

            set
            {
                ViewState["AutoLoadItems"] = value;
            }
        }

        /// <summary>
        /// Gets or sets the data view entity type identifier.
        /// </summary>
        /// <value>
        /// The data view entity type identifier.
        /// </value>
        public int? EntityTypeId
        {
            get
            {
                return ViewState["EntityTypeId"] as int?;
            }

            set
            {
                ViewState["EntityTypeId"] = value;
                if ( AutoLoadItems )
                {
                    LoadDropDownItems();
                }
            }
        }

        /// <summary>
        /// Gets or sets an optional list of category guids to limit dataviews by
        /// </summary>
        /// <value>
        /// The category guids.
        /// </value>
        public List<Guid> CategoryGuids
        {
            get
            {
                return ViewState["CategoryGuids"] as List<Guid>;
            }

            set
            {
                ViewState["CategoryGuids"] = value;

                if ( AutoLoadItems )
                {
                    LoadDropDownItems();
                }
            }
        }

        /// <summary>
        /// Loads the drop down items.
        /// </summary>
        public void LoadDropDownItems()
        {
            this.Items.Clear();

            if ( EntityTypeId.HasValue )
            {
                // add Empty option first
                this.Items.Add( new ListItem() );

                // Get 
                var categoryGuids = CategoryGuids ?? new List<Guid>();

                using ( var rockContext = new RockContext() )
                {
                    var allEntityFilters = new DataViewFilterService( rockContext )
                        .Queryable().AsNoTracking()
                        .Where( f => f.EntityTypeId == EntityTypeId )
                        .ToList();
                        
                    foreach ( var dataView in new DataViewService( rockContext )
                        .GetByEntityTypeId( EntityTypeId.Value )
                        .Include( "DataViewFilter" )
                        .AsNoTracking() )
                    {
                        var category = dataView.CategoryId.HasValue ? CategoryCache.Get( dataView.CategoryId.Value ) : null;
                        if ( !categoryGuids.Any() || ( category != null && categoryGuids.Contains( category.Guid ) ) )
                        { 
                            var currentPerson = HttpContext.Current.Items["CurrentPerson"] as Person;
                            if ( dataView.IsAuthorized( Authorization.VIEW, currentPerson ) &&
                                dataView.DataViewFilter.IsAuthorized( Authorization.VIEW, currentPerson, allEntityFilters ) )
                            {
                                this.Items.Add( new ListItem( dataView.Name, dataView.Id.ToString() ) );
                            }
                        }
                    }
                }
            }
        }
    }
}

File: Rock.JavaScript.Obsidian/Framework/Controls/dataViewPicker.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//

import { Guid } from "@Obsidian/Types";
import { useSecurityGrantToken } from "@Obsidian/Utility/block";
import { standardAsyncPickerProps, updateRefValue } from "@Obsidian/Utility/component";
import { ListItemBag } from "@Obsidian/ViewModels/Utility/listItemBag";
import { defineComponent, PropType, ref, watch } from "vue";
import { DataViewTreeItemProvider } from "@Obsidian/Utility/treeItemProviders";
import TreeItemPicker from "./treeItemPicker";

export default defineComponent({
    name: "DataViewPicker",

    components: {
        TreeItemPicker
    },

    props: {
        modelValue: {
            type: Object as PropType<ListItemBag | ListItemBag[] | null>,
            required: false
        },

        entityTypeGuid: {
            type: String as PropType<Guid>,
            required: false
        },

        ...standardAsyncPickerProps
    },

    emits: {
        "update:modelValue": (_value: ListItemBag | ListItemBag[] | null) => true
    },

    setup(props, { emit }) {
        // #region Values

        const internalValue = ref(props.modelValue ?? null);
        const securityGrantToken = useSecurityGrantToken();

        const itemProvider = ref(new DataViewTreeItemProvider());
        itemProvider.value.entityTypeGuid = props.entityTypeGuid;
        itemProvider.value.securityGrantToken = securityGrantToken.value;

        // #endregion

        // #region Watchers

        // Keep security token up to date, but don't need refetch data
        watch(securityGrantToken, () => {
            itemProvider.value.securityGrantToken = securityGrantToken.value;
        });

        // When this changes, we need to refetch the data, so reset the whole itemProvider
        watch(() => props.entityTypeGuid, () => {
            const oldProvider = itemProvider.value;
            const newProvider = new DataViewTreeItemProvider();

            // copy old provider's properties
            newProvider.securityGrantToken = oldProvider.securityGrantToken;
            // Use new value
            newProvider.entityTypeGuid = props.entityTypeGuid;

            // Set the provider to the new one
            itemProvider.value = newProvider;
        });

        watch(internalValue, () => {
            emit("update:modelValue", internalValue.value);
        });

        watch(() => props.modelValue, () => {
            updateRefValue(internalValue, props.modelValue ?? null);
        });

        // #endregion

        return {
            internalValue,
            itemProvider
        };
    },
    template: `
<TreeItemPicker v-model="internalValue"
    formGroupClasses="category-picker"
    iconCssClass="fa fa-filter"
    :provider="itemProvider"
    :multiple="multiple"
    disableFolderSelection
/>
`
});
